### PR TITLE
closes #332, type parser is strengthened with Scalacheck

### LINKED ISSUE #332: [BUG] Syntactically invalid type annotations on operators results in uncaught MatchError
I'll fix this.  Just creating an issue to document and organize my work.

---

**Describe the bug**

Etc type checker parser crashes on some syntactically incorrect type signatures.

This turned up when working on #329 

**To Reproduce**

Any operator annotated with a syntactically incorrect target type seems to suffice for reproduction E.g.,

```tlaplus
move(p, d) == "() => <<))" ## 
    ...
```

**Expected behavior**

We should get a helpful error indicating, at least, the location at which the parsing error arose. 

**Log files**

From a failing integration test:

```
--- /home/sf/Sync/informal-systems/apalache/apalache-core/target/test/tla/cli-integration-tests.md.corrected	2020-11-17 16:49:19.161046589 -0500
+++ /home/sf/Sync/informal-systems/apalache/apalache-core/test/tla/cli-integration-tests.md	2020-11-17 16:17:16.194558624 -0500
@@ -682,38 +682,13 @@
 ...
 PASS #1: TypeChecker
  > running ETC: Embarrassingly simple Type Checker
    ^_^
-Unhandled exception
-scala.MatchError: (List(<<Int, Int>>, <<Int, Int>>)~None) (of class scala.util.parsing.combinator.Parsers$$tilde)
-	at at.forsyte.apalache.tla.typecheck.parser.DefaultType1Parser$.$anonfun$typeExpr$5(DefaultType1Parser.scala:55)
-	at scala.util.parsing.combinator.Parsers$Success.map(Parsers.scala:140)
-	at scala.util.parsing.combinator.Parsers$Success.map(Parsers.scala:139)
-	at scala.util.parsing.combinator.Parsers$Parser.$anonfun$map$1(Parsers.scala:235)
-	at scala.util.parsing.combinator.Parsers$$anon$1.apply(Parsers.scala:215)
-	at scala.util.parsing.combinator.Parsers$$anon$5.apply(Parsers.scala:919)
-	at scala.util.parsing.combinator.Parsers$Parser.$anonfun$map$1(Parsers.scala:235)
-	at scala.util.parsing.combinator.Parsers$$anon$1.apply(Parsers.scala:215)
-	at at.forsyte.apalache.tla.typecheck.parser.DefaultType1Parser$.apply(DefaultType1Parser.scala:42)
-	at at.forsyte.apalache.tla.typecheck.parser.DefaultType1Parser$.apply(DefaultType1Parser.scala:33)
-	at at.forsyte.apalache.tla.typecheck.etc.ToEtcExpr.apply(ToEtcExpr.scala:45)
-	at at.forsyte.apalache.tla.typecheck.TypeCheckerTool.$anonfun$check$1(TypeCheckerTool.scala:24)
-	at scala.collection.immutable.List.$anonfun$foldRight$1(List.scala:447)
-	at scala.collection.immutable.List.foldRight(List.scala:91)
-	at at.forsyte.apalache.tla.typecheck.TypeCheckerTool.check(TypeCheckerTool.scala:23)
-	at at.forsyte.apalache.tla.typecheck.passes.EtcTypeCheckerPassImpl.execute(EtcTypeCheckerPassImpl.scala:31)
-	at at.forsyte.apalache.infra.passes.PassChainExecutor.exec$1(PassChainExecutor.scala:23)
-	at at.forsyte.apalache.infra.passes.PassChainExecutor.run(PassChainExecutor.scala:38)
-	at at.forsyte.apalache.tla.Tool$.runTypeCheck(Tool.scala:168)
-	at at.forsyte.apalache.tla.Tool$.$anonfun$run$3(Tool.scala:88)
-	at at.forsyte.apalache.tla.Tool$.$anonfun$run$3$adapted(Tool.scala:88)
-	at at.forsyte.apalache.tla.Tool$.handleExceptions(Tool.scala:216)
-	at at.forsyte.apalache.tla.Tool$.run(Tool.scala:88)
-	at at.forsyte.apalache.tla.Tool$.main(Tool.scala:43)
-	at at.forsyte.apalache.tla.Tool.main(Tool.scala)
-It took me 0 days  0 hours  0 min  1 sec
-Total time: 1.557 sec
-EXITCODE: ERROR (99)
+  =^_^=
+PASS #2: Terminal
+Type checker [OK]
+...
+EXITCODE: OK
 ```

### PR DESCRIPTION
**Pull request by konnov**
This PR fixes the bug reported in #332 and a few small inconsistencies that were found with Scalacheck.

- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] ~Documentation added for any new functionality~
- [x] Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality

### CODE REVIEW

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/TestDefaultType1Parser.scala [hunk 1]
```diff
@@ -122,4 +137,46 @@ class TestDefaultType1Parser  extends FunSuite {
...
   }
 
-}
\ No newline at end of file
+  test("vc") {
+    // found by scalacheck
+    assertThrows[Type1ParseError](DefaultType1Parser("vc"))
+  }
+
+  test("arbitrary inputs") {
```
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/TestDefaultType1Parser.scala:145]: This test name is not very helpful.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/TlaType1Gen.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  def genConst: Gen[TlaType1] =
+    for {
+      str <- identifier
+    } yield ConstT1(str.toUpperCase())
+
+  def genVar: Gen[TlaType1] =
+    for {
+      i <- arbitrary[Int]
+      // produce an absolute value. Note that Math.abs(Integer.MIN_VALUE) = Integer.MIN_VALUE, so use max(0, abs(i)).
+    } yield VarT1(Math.max(0, Math.abs(i)))
```
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/TlaType1Gen.scala:18]: Should this use `Gen.posNum`? https://github.com/typelevel/scalacheck/blob/8645105297ff26512d002e87d413717212892977/src/main/scala/org/scalacheck/Gen.scala#L1243 (Usage here: https://github.com/typelevel/scalacheck/blob/2224f8506ca8ab079e3a0c4d18d30db7a0395e70/jvm/src/test/scala/org/scalacheck/PropertyFilterSpecification.scala#L16)

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/TestDefaultType1Parser.scala [hunk 3]
```diff
@@ -122,4 +137,46 @@ class TestDefaultType1Parser  extends FunSuite {
...
+    check({
+      forAll(alphaStr) { str =>
+        try {
+          DefaultType1Parser(str)
+          passed
+        } catch {
+          case _: Type1ParseError =>
+            passed
```
[the remark this answers is not in the document]
Comment (konnov) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecheck/TestDefaultType1Parser.scala:153]: This test checks that the parser does not crash on arbitrary input. It either parses the text and return an annotation, or throws a parser error, which is expected behavior.

**Review by shonfeder — APPROVED** [review]
LGTM, tho I have one suggestion.

**Pull request merged**